src/components/treeView/createEntityNode.tsx

```diff
@@ -0,0 +1,169 @@
+import * as React from 'react';
+
+import { NodeRenderStrategy } from './nodeRenderStrategy';
+import { LeafNode, LeafNodeProps } from './leafNode';
+import { InnerGlobals } from '../../props/globalProps';
+import { NameValidator } from '../../nameValidator';
+import { Constants } from '../../constants';
+
+export interface CreateEntityNodeProps extends InnerGlobals {
+	tabbable: boolean;
+	level: number;
+	notStartedRenderStrategy: (onClick: () => void) => NodeRenderStrategy;
+	inputRenderStrategy:
+		(inputValue: string,
+			onEnter: () => void,
+			onInputChange: (evt: React.ChangeEvent<HTMLInputElement>) => void,
+			setInputRefAndFocus: (node: HTMLInputElement) => void) => NodeRenderStrategy;
+	createErrorRenderStrategy: (inputValue: string, onInputChange: (evt: React.ChangeEvent<HTMLInputElement>) => void) => NodeRenderStrategy;
+	inProgressRenderStrategy: (inputValue: string) => NodeRenderStrategy;
+	createEntity: (name: string) => Promise<void>;
+}
+
+export interface CreateEntityNodeState {
+	status: 'NotStarted' | 'Input' | 'CreateError' | 'InProgress';
+	nameInputValue: string;
+}
+
+/**
+ * Functional component for creating a particular entity e.g., notebook. UI logic
+ * should be passed in as props.
+ */
+export class CreateEntityNode extends React.Component<CreateEntityNodeProps, CreateEntityNodeState> {
+	private wrapperRef: HTMLElement;
+	private inputRef: HTMLInputElement;
+	private componentIsMounted: boolean;
+
+	constructor() {
+		super();
+		this.state = this.defaultState();
+
+		this.setWrapperRef = this.setWrapperRef.bind(this);
+		this.setInputRefAndFocus = this.setInputRefAndFocus.bind(this);
+		this.handleClickOutside = this.handleClickOutside.bind(this);
+		this.handleEnterInput = this.handleEnterInput.bind(this);
+		this.resetAndFocus = this.resetAndFocus.bind(this);
+		this.onInputChange = this.onInputChange.bind(this);
+		this.onClick = this.onClick.bind(this);
+	}
+
+	componentDidMount() {
+		document.addEventListener('mousedown', this.handleClickOutside);
+		this.componentIsMounted = true;
+	}
+
+	componentWillUnmount() {
+		document.removeEventListener('mousedown', this.handleClickOutside);
+		this.componentIsMounted = false;
+	}
+
+	private handleClickOutside(event) {
+		if (this.state.status === 'Input' && this.wrapperRef && !this.wrapperRef.contains(event.target)) {
+			this.handleEnterInput();
+		}
+	}
+
+	private handleEnterInput() {
+		if (NameValidator.validateNotebookName(this.state.nameInputValue)) {
+			// Don't fire off request or change state if user attempts to submit invalid name
+			return;
+		}
+
+		// Check for whitespace
+		if (this.state.nameInputValue && this.state.nameInputValue.trim()) {
+			this.setState({ status: 'InProgress' });
+
+			this.props.createEntity(this.state.nameInputValue).then(() => {
+				if (this.componentIsMounted) {
+					this.setState(this.defaultState());
+				}
+			}).catch((error) => {
+				this.setState({
+					status: 'CreateError'
```

Comment:
Should we also set an object here describing the error so that we could show it in the UI later?

Reply:
It's ok to not have this diagnostic information in the initial iteration, but I think if we ship this, we need a way for customers to give us diagnostic information, otherwise it'll be complicated to debug issues.

Reply:
I've added a TODO, but in the meantime I won't set the error object in state since we don't use it